src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -413,8 +413,11 @@ AcquireExecutorMultiShardLocks(List *taskList)
 		 * If we are dealing with a partition we are also taking locks on parent table
 		 * to prevent deadlocks on concurrent operations on a partition and its parent.
 		 */
-		LockParentShardResourceIfPartition(task->anchorShardId, lockMode);
-		LockShardResource(task->anchorShardId, lockMode);
+		if (task->anchorShardId != INVALID_SHARD_ID)
```

Comment:
The function name is `AcquireExecutorMultiShardLocks ` and we've `task->anchorShardId != INVALID_SHARD_ID`. This might be confusing and could lead to future bugs via other callers. 

Can't we add a  check in the adaptive executor `AcquireExecutorShardLocksForExecution` and return if the tasks doesn't have valid shardIds?

Reply:
To have this check earlier is complicated for the following reasons:
 - the list is passed, so there might be some tasks that are anchored to a shard and some that are not
 - there are multiple call sites to this, so the check on the list needs to be done in a loop, multiple times.

instead I can move the check up in this look and `continue;` the loop if there is no `anchorShardId` on the task it is currently trying to take a lock on

Reply:
Looking at the code again, there is a codepath below that would be skipped as well with a continue. This seems to be closely related to shards as well and I assume this would always be an empty list when not anchored to a shard.